### PR TITLE
Exo-player two facade and forwarders

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
`ExoPlayerForwarder` contained all of the events that we listen to from `ExoPlayer` and we created `Forwarders` that adapted the `ExoPlayer` events to our `no-player` events. Some of these `Forwarders` contained state. `Facade` and `Imple` were mixing responsibilities with no clear boundary.

## Solution
- Remove the catch all `Forwarder` and implement specific forwarders for the events we want to listen to.
- Create a generic forwarder that has as collaborators the specific forwarder implementations.
- Remove the `Facade` from exo-player because the code is small enough to be encapsulated in one class with collaborators.

### Test(s) added 
No tests atm, we hope to branch off and have these covered soon.

### Screenshots
No UI changes.

### Paired with 
@Dorvaryn and @jackSzm

### CODE REVIEW

File: core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerEventForwarder.java [hunk 1]
```diff
@@ -1,122 +0,0 @@
...
-import com.google.android.exoplayer2.PlaybackParameters;
-import com.google.android.exoplayer2.Timeline;
-import com.google.android.exoplayer2.decoder.DecoderCounters;
-import com.google.android.exoplayer2.source.TrackGroupArray;
-import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
-import com.google.android.exoplayer2.upstream.DataSpec;
-
-import java.io.IOException;
-
-abstract class ExoPlayerEventForwarder implements ExoPlayerForwarder {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerEventForwarder.java]: 🤢  we implemented all events on this one forwarder. We now have dedicated forwarders and a generic `forwarder` that has the specific forwarders as collaborators.

File: core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java [hunk 2]
```diff
@@ -13,74 +22,93 @@
 import com.novoda.noplayer.VideoContainer;
 import com.novoda.noplayer.VideoDuration;
 import com.novoda.noplayer.VideoPosition;
+import com.novoda.noplayer.exoplayer.forwarder.ExoPlayerForwarder;
 import com.novoda.noplayer.player.PlayerInformation;
 
 import java.util.Collections;
 import java.util.List;
 
 public class ExoPlayerTwoImpl extends PlayerListenersHolder implements Player {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java]: Facade became a wrapper during out refactorings so quite a few of the methods here are just inlines.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BitrateForwarder.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.google.android.exoplayer2.C;
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.source.AdaptiveMediaSourceEventListener;
+import com.google.android.exoplayer2.upstream.DataSpec;
+import com.novoda.noplayer.listeners.BitrateChangedListeners;
+
+import java.io.IOException;
+
+class BitrateForwarder implements AdaptiveMediaSourceEventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BitrateForwarder.java]: No longer extending from the abstract `ExoPlayerEventForwarder`, instead, we map directly from the `ExoPlayer` events to the `no-player` events.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BitrateForwarder.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            bitrateChangedListeners.onBitrateChanged(audioBitrate, videoBitrate);
+        } else if (trackType == C.TRACK_TYPE_AUDIO) {
+            audioBitrate = Bitrate.fromBitsPerSecond(trackFormat.bitrate);
+            bitrateChangedListeners.onBitrateChanged(audioBitrate, videoBitrate);
+        }
+    }
+
+    @Override
+    public void onLoadStarted(DataSpec dataSpec, int dataType, int trackType, Format trackFormat, int trackSelectionReason, Object trackSelectionData, long mediaStartTimeMs, long mediaEndTimeMs, long elapsedRealtimeMs) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BitrateForwarder.java]: These `TODOs` were hidden in the abstract `ExoPlayerEventForwarder` class, not anymore! We should figure out what to do with these!

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BufferStateForwarder.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.google.android.exoplayer2.ExoPlaybackException;
+import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.PlaybackParameters;
+import com.google.android.exoplayer2.Timeline;
+import com.google.android.exoplayer2.source.TrackGroupArray;
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
+import com.novoda.noplayer.listeners.BufferStateListeners;
+
+class BufferStateForwarder implements ExoPlayer.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BufferStateForwarder.java]: Same here, not extending from the abstract `ExoPlayerEventForwarder` instead implementing the interface from `ExoPlayer` that we care about.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BufferStateForwarder.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        if (playbackState == ExoPlayer.STATE_BUFFERING) {
+            bufferStateListeners.onBufferStarted();
+        } else if (playbackState == ExoPlayer.STATE_READY) {
+            bufferStateListeners.onBufferCompleted();
+        }
+    }
+
+    @Override
+    public void onTimelineChanged(Timeline timeline, Object manifest) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/BufferStateForwarder.java]: Same here, we need to figure out how we should map these from `ExoPlayer` to `no-player`

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/EventInfoForwarder.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.PlaybackParameters;
+import com.google.android.exoplayer2.Timeline;
+import com.google.android.exoplayer2.source.TrackGroupArray;
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
+import com.novoda.noplayer.listeners.InfoListeners;
+
+import java.util.HashMap;
+
+class EventInfoForwarder implements ExoPlayer.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/EventInfoForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/EventInfoForwarder.java]: At some point this will likely contain more implementations, being an `info` forwarder, it is the most generic reporter we have, mainly useful for debugging. 🤷‍♂️

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerExtractorMediaSourceListener.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.novoda.noplayer.exoplayer.forwarder;
+
+import com.google.android.exoplayer2.source.ExtractorMediaSource;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+class ExoPlayerExtractorMediaSourceListener implements ExtractorMediaSource.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerExtractorMediaSourceListener.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerForwarder.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import com.novoda.noplayer.PlayerState;
+import com.novoda.noplayer.listeners.BitrateChangedListeners;
+import com.novoda.noplayer.listeners.BufferStateListeners;
+import com.novoda.noplayer.listeners.CompletionListeners;
+import com.novoda.noplayer.listeners.ErrorListeners;
+import com.novoda.noplayer.listeners.InfoListeners;
+import com.novoda.noplayer.listeners.PreparedListeners;
+import com.novoda.noplayer.listeners.VideoSizeChangedListeners;
+
+public class ExoPlayerForwarder {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerForwarder.java]: Our generic forwarder that contains all of the `events` that we are interested in from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerVideoRendererEventListener.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import android.view.Surface;
+
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.decoder.DecoderCounters;
+import com.google.android.exoplayer2.video.VideoRendererEventListener;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+class ExoPlayerVideoRendererEventListener implements VideoRendererEventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerVideoRendererEventListener.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExtractorInfoForwarder.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
+package com.novoda.noplayer.exoplayer.forwarder;
+
+import com.google.android.exoplayer2.source.ExtractorMediaSource;
+import com.novoda.noplayer.listeners.InfoListeners;
+
+import java.io.IOException;
+import java.util.HashMap;
+
+class ExtractorInfoForwarder implements ExtractorMediaSource.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExtractorInfoForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExtractorInfoForwarder.java]: To avoid huge classes we have an `InfoForwarder` for each of the `ExoPlayer` interfaces that we are interested in monitoring.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/MediaSourceInfoForwarder.java [hunk 12]
```diff
@@ -1,21 +1,18 @@
...
-import com.google.android.exoplayer2.decoder.DecoderCounters;
+import com.google.android.exoplayer2.source.AdaptiveMediaSourceEventListener;
 import com.google.android.exoplayer2.upstream.DataSpec;
 import com.novoda.noplayer.listeners.InfoListeners;
 
 import java.io.IOException;
 import java.util.HashMap;
 
-class InfoForwarder extends ExoPlayerEventForwarder {
+class MediaSourceInfoForwarder implements AdaptiveMediaSourceEventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/MediaSourceInfoForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/MediaSourceOnErrorForwarder.java [hunk 13]
```diff
@@ -1,32 +1,25 @@
...
-import com.google.android.exoplayer2.ExoPlaybackException;
+import com.google.android.exoplayer2.source.ExtractorMediaSource;
 import com.novoda.noplayer.Player;
 import com.novoda.noplayer.exoplayer.playererror.MediaSourceError;
 import com.novoda.noplayer.listeners.ErrorListeners;
 
 import java.io.IOException;
 
-class OnErrorForwarder extends ExoPlayerEventForwarder {
+class MediaSourceOnErrorForwarder implements ExtractorMediaSource.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/MediaSourceOnErrorForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.google.android.exoplayer2.ExoPlaybackException;
+import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.PlaybackParameters;
+import com.google.android.exoplayer2.Timeline;
+import com.google.android.exoplayer2.source.TrackGroupArray;
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
+import com.novoda.noplayer.listeners.CompletionListeners;
+
+class OnCompletionForwarder implements ExoPlayer.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
+        if (playbackState == ExoPlayer.STATE_ENDED) {
+            completionListeners.onCompletion();
+        }
+    }
+
+    @Override
+    public void onTimelineChanged(Timeline timeline, Object manifest) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java]: Again, we need to figure out the mapping from `ExoPlayer` to `no-player` for these additional events.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.android.exoplayer2.ExoPlaybackException;
+import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.PlaybackParameters;
+import com.google.android.exoplayer2.Timeline;
+import com.google.android.exoplayer2.source.TrackGroupArray;
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
+import com.novoda.noplayer.PlayerState;
+import com.novoda.noplayer.listeners.PreparedListeners;
+
+class OnPrepareForwarder implements ExoPlayer.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    private boolean isReady(int playbackState) {
+        return playbackState == ExoPlayer.STATE_READY;
+    }
+
+    @Override
+    public void onTimelineChanged(Timeline timeline, Object manifest) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java]: Once again, again. We need to figure out the mapping between `ExoPlayer` and `no-player` for these additional events 😄

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+import com.google.android.exoplayer2.ExoPlaybackException;
+import com.google.android.exoplayer2.ExoPlayer;
+import com.google.android.exoplayer2.PlaybackParameters;
+import com.google.android.exoplayer2.Timeline;
+import com.google.android.exoplayer2.source.TrackGroupArray;
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
+import com.novoda.noplayer.Player;
+import com.novoda.noplayer.listeners.ErrorListeners;
+
+class PlayerOnErrorForwarder implements ExoPlayer.EventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    public void onPlayerError(ExoPlaybackException error) {
+        Player.PlayerError playerError = ExoPlayerErrorFactory.errorFor(error);
+        errorListeners.onError(player, playerError);
+    }
+
+    @Override
+    public void onTimelineChanged(Timeline timeline, Object manifest) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java]: AND AGAIN! 😄 🎉  We need to figure out the mapping between `exoplayer` and `no-player`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoRendererInfoForwarder.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+import android.view.Surface;
+
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.decoder.DecoderCounters;
+import com.google.android.exoplayer2.video.VideoRendererEventListener;
+import com.novoda.noplayer.listeners.InfoListeners;
+
+import java.util.HashMap;
+
+class VideoRendererInfoForwarder implements VideoRendererEventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoRendererInfoForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoRendererInfoForwarder.java]: There are many info forwarders and each implements a single interface from `ExoPlayer` and are added to the `generic forwarder`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoSizeChangedForwarder.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
+package com.novoda.noplayer.exoplayer.forwarder;
+
+import android.view.Surface;
+
+import com.google.android.exoplayer2.Format;
+import com.google.android.exoplayer2.decoder.DecoderCounters;
+import com.google.android.exoplayer2.video.VideoRendererEventListener;
+import com.novoda.noplayer.listeners.VideoSizeChangedListeners;
+
+class VideoSizeChangedForwarder implements VideoRendererEventListener {
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoSizeChangedForwarder.java]: Same as others above. We no longer extend from `ExoPlayerEventForwarder` but instead implement the interface we care about directly from `ExoPlayer`.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoSizeChangedForwarder.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public void onVideoSizeChanged(int width, int height, int unappliedRotationDegrees, float pixelWidthHeightRatio) {
+        videoSizeChangedListeners.onVideoSizeChanged(width, height, unappliedRotationDegrees, pixelWidthHeightRatio);
+    }
+
+    @Override
+    public void onVideoEnabled(DecoderCounters counters) {
+        //TODO should we send ?
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/VideoSizeChangedForwarder.java]: OOOOoooo more! We need to figure out the mapping between `ExoPlayer` and `no-player`.

File: core/src/main/java/com/novoda/noplayer/listeners/PreparedListeners.java [hunk 23]
```diff
@@ -10,6 +10,8 @@
 
     private final Set<Player.PreparedListener> listeners = new CopyOnWriteArraySet<>();
 
+    private boolean hasPrepared = false;
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/listeners/PreparedListeners.java]: We moved the state up a level because it made more sense here than in each of the individual listeners.

File: demo/src/main/java/com/novoda/demo/MainActivity.java [hunk 24]
```diff
@@ -30,7 +30,7 @@ protected void onCreate(Bundle savedInstanceState) {
     @Override
     protected void onStart() {
         super.onStart();
-        player = new PlayerFactory(this, PrioritisedPlayers.prioritiseMediaPlayer()).create();
+        player = new PlayerFactory(this, PrioritisedPlayers.prioritiseExoPlayer()).create();
```
Comment (Mecharyry) [inline demo/src/main/java/com/novoda/demo/MainActivity.java]: 🤷‍♂️ because we are working on `ExoPlayer` 👅

Comment (rock3r) [inline demo/src/main/java/com/novoda/demo/MainActivity.java]: Leave a TODO maybe?

File: core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java [hunk 25]
```diff
@@ -112,27 +142,39 @@ public void reset() {
...
     @Override
     public void loadVideo(Uri uri, ContentType contentType) {
-        prepareForwarder.reset();
+        getPreparedListeners().reset();
         showContainer();
-        facade.prepare(uri, contentType);
+        exoPlayer.addListener(forwarder.exoPlayerEventListener());
+        exoPlayer.setVideoDebugListener(forwarder.videoRendererEventListener());
+
+        exoPlayer.setPlayWhenReady(true);
```
Comment (ouchadam) [inline core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java]: what's the reason for this change? it'll break symmetry with the media player as the we're relying on 

`loadVideo -> onPrepared -> play`

Comment (jackSzm) [inline core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java]: I think we no longer need onPrepared step in exoplayer2.
But if that's the case in mediaPlayer then probably we would need to bring it back.

File: demo/src/main/java/com/novoda/demo/MainActivity.java (hunk 24 above)
Comment (Mecharyry) [inline demo/src/main/java/com/novoda/demo/MainActivity.java]: Yea, we should have this exposed in the UI really.

File: core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java [hunk 26]
```diff
@@ -13,74 +22,95 @@
...
+        return new ExoPlayerTwoImpl(exoPlayer, mediaSourceFactory, new ExoPlayerForwarder());
     }
 
-    private ExoPlayerTwoImpl(ExoPlayerTwoFacade facade) {
-        this.facade = facade;
+    private ExoPlayerTwoImpl(SimpleExoPlayer exoPlayer, MediaSourceFactory mediaSourceFactory, ExoPlayerForwarder exoPlayerForwarder) {
+        this.exoPlayer = exoPlayer;
+        this.mediaSourceFactory = mediaSourceFactory;
+        forwarder = exoPlayerForwarder;
         Heart.Heartbeat<Player> onHeartbeat = new Heart.Heartbeat<>(getHeartbeatCallbacks(), this);
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/ExoPlayerTwoImpl.java]: I think we have done this in a follow-up branch that we are waiting to open 😆  but yes we should try and move all of these out of the constructor.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public void onLoadingChanged(boolean isLoading) {
+        //TODO should we send ?
+    }
+
+    @Override
+    public void onPlayerError(ExoPlaybackException error) {
+        //Sent by ErrorForwarder
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java]: Ah. This particular event is handled by the `errorForwarder` and should not be handled by this `Forwarder`. This is because it would be a bit strange to notify the `no-player` complete listeners of an error, when that is handled by the `no-player` error listeners. 

This is a small problem with using the interfaces we are interested in from `ExoPlayer` it leads to explicit no-ops in the specific implementations.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public void onLoadingChanged(boolean isLoading) {
+        //TODO should we send ?
+    }
+
+    @Override
+    public void onPlayerError(ExoPlaybackException error) {
+        //Sent by ErrorForwarder
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnPrepareForwarder.java]: Same as `onCompletionListener` response 👅

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Override
+    public void onLoadingChanged(boolean isLoading) {
+        //TODO should we send ?
+    }
+
+    @Override
+    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
+        //Handled by OnPrepared and OnCompletion forwarders
```
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java]: Yea, it is because `ExoPlayer.EventListener` is too broad a listener 😢  but we decided it was better than having an abstract class. Better to be explicit about these in the implementations then to hide it in the abstraction.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java (hunk 27 above)
Comment (rock3r) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/OnCompletionForwarder.java]: Figured that was the case, ok :)

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java (hunk 29 above)
Comment (rock3r) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/PlayerOnErrorForwarder.java]: Yep agreed, I'm ok with that
